### ISSUE TITLE
Is there a way to toggle on/off an include/Shader/TextureOverride?

### ISSUE THREAD

**Issue by biast12** [bo3b/3Dmigoto#159]
ok so i'm asking this because i'm using "include" to call other .ini files to inject ShaderOverride/TextureOverride, and i kinda want to toggle them on an off instead of needing to do that manually every time i wanna change something, an example of that would be the UI, i have tried to just make it a key, but it doesn't do anything, is this an .ini issue or an 3Dmigoto issue? and would there be a chance that it could either be added for the include or for the Shader/Texture injecting system

**Comment by Nintynuts** [comment]
I don't believe it's possible to toggle an entire `include` since that's done statically at load time, but within your shader and texture overrides that you want to toggle off you can surround the command list with an `if` statement using a custom variable (which is evaluated every frame), and you can toggle the value of that variable with whichever key you want by setting up a key binding. If I understand what you're trying to do, that should give the desired effect.

Can @bo3b or @DarkStarSword confirm and close if @biast12 is happy with that answer?

**Comment by DarkStarSword** [comment]
Ninjad ;-)
Yeah, what Nintynuts said - config load time is too early to process a toggle, that has to be implemented at runtime in a command list and/or shader... Here's what I had started typing up before getting dinner in case it's of any interest:

Do I understand right that you want a quick toggle to choose to exclude the included ini file in it's entirety? This isn't really possible in the way you are thinking as that decision would have to be made when the configuration is [re]loaded, and that is too early to be able to evaluate dynamic expressions (e.g. variables that a key binding might set don't exist yet because they haven't been parsed or initialised by that point, and cannot be because - chicken, meet egg).

There *is* an existing facility to conditionally exclude ini files ([example](https://github.com/DarkStarSword/3d-fixes/blob/master/DOA6/Mods/3d-vision/stereo%20correction.ini#L1)), but it only works with statically evaluatable expressions - in practice is essentially just used for "exclude this ini file on systems that do / do not support 3D Vision and/or SLI" (and even then only the availability of 3D Vision can be statically evaluated - it can't even make the determination on whether 3D Vision is currently enabled or not as that can be changed at runtime and is therefore not statically evaluatable on systems that can support it). You could use this to set a simple static expression at the top of the ini file such as `condition = 1` / `condition = 0`, but you still would have to edit the file and reload config to change that (maybe that is easier than renaming the file if you're using recursive include, but probably doesn't help much if you have manually listed the included file since it's the same amount of effort as commenting/uncommenting the include line).

In general, if you want to toggle some functionality at runtime (and especially if you intend users to be able to do this) you need to put the check either in one of the ini sections that support command lists (ShaderOverride + TextureOverride both do, but with the caveat that both those sections have several non-command list options as well that won't be affected by such a toggle - [list of ShaderOverride exceptions](https://github.com/bo3b/3Dmigoto/blob/master/DirectX11/IniHandler.cpp#L2227), [list of TextureOverride exceptions](https://github.com/bo3b/3Dmigoto/blob/master/DirectX11/IniHandler.cpp#L2227)), or in the shaders themselves.

There's plenty of examples of both of these, for example the [$costume_mods variable](https://github.com/bo3b/3Dmigoto/blob/master/DirectX11/IniHandler.cpp#L2227) in DOAXVV for the former, and ... if you need an example of doing it in a shader ask - I'm sure I have plenty of examples throughout my 3d-fixes repo, but the ones I can think of off the top of my head are either overly complicated to follow, or are using older techniques that we may have better ways to do now.

**Comment by DarkStarSword** [comment]
Another option if it's just to make things easier for yourself when creating a mod might be to use something external to 3DMigoto, for example you could write a bat file to switch the ini files in and out, or even hack to gether a quick n' dirty AHK script to do it on hotkey.

**Comment by DarkStarSword** [comment]
No, adding such a feature is going to cause more pain, bugs and edge cases you haven't thought of than it's worth (quick example - what do you suppose would happen if that [Key1] section is itself inside an included file?), and considering that we have good alternatives to toggle HUDs and other things that have been in use for years this feature request is a NAK.
